### ISSUE TITLE
[Feature Request] Add an option to throw an error after a configured amount of operations or time.

### ISSUE THREAD

**Issue by TapuCosmo** [GreenImp/rpg-dice-roller#137]
**Is your feature request related to a problem? Please describe.**
Some input strings, such as `d6!>0`, can cause an infinite loop. Therefore, I need to manually check input strings, but I do not know all of the possible strings that will cause the script to get stuck or run for too long.

**Describe the solution you'd like**
An option to make the dice rolls stop and throw an error after a configured amount of operations or time would allow the caller of the dice roll to regain control. For example, a user can specify a limit of 100 operations to force the dice roller to stop after 100 operations, or a time limit of 25 to prevent the dice roller from running for more than 25ms on any single roll. Time limits can check `performance.now()` from perf_hooks for increased resolution.

**Describe alternatives you've considered**
Ideally, the dice roller should throw errors from the beginning for syntax that would lead to an infinite loop, but it is difficult to ensure that all possible infinite loops are caught, and this alternative would also be unable to detect syntax that do not get stuck in a loop, but take up a lot of time.

**Additional context**
Use Case: I use this module in a Discord bot, so having the event loop get stuck in the dice roll is extremely bad, since people can exploit it to cause a denial of service. Currently, I am regex matching for known problematic syntax, but it is possible that someone might eventually discover a new input string that I have not filtered to crash the process.

**Comment by GreenImp** [comment]
That's a very interesting conundrum, and it definitely needs a solution.

I'll have to think about a good way around this. I think throwing an error for things like a compare point that will always return true is a god idea, it's just determining how that can be done.

The option of a timeout is an interesting one as well. I think this needs some proper investigation to see determine what notation can acuse such loops/

Do you have other examples that can cause infinite loops?

**Comment by TapuCosmo** [comment]
I have `!>=0`, `!p>=0`, `!p>0`, and a check for 5 digit numbers (I don't remember why exactly I added this check, but 5 digit numbers used in a certain way caused the dice roll to be very slow).

**Comment by TapuCosmo** [comment]
I think the reason for the 5-digit limit was to prevent things like `9999999d20`.

**Comment by Yernemm** [comment]
Any progress on this? I am also trying to run this on a bot and for safety I ended up completely banning certain operations like '>', '<', '!', 'r', and some others, as well as a maximum digit limit as suggested above. 

Checking for specifics is dangerous since I found things like `!>=0` would also trigger a long loop on negative inputs, i.e. `!>=-` so I think a better work-around right now for those looking to use this in a bot would be completely disabling any operations that may lead to a long loop. My current banned list is `['>', '<', '!', "r", "exp", "log", "max", "min", "pow", "sign"]`. Some of these may be fine but I'm being pretty cautious here so I banned some which may seem dangerous, though untested.

It would be helpful to maybe have options for ignoring certain dangerous modifiers or indeed have it terminate if it takes too long. Perhaps writing an asynchronous version would be helpful.

**Comment by GreenImp** [comment]
Sorry, no progress yet. It's been earmarked for the 4.3.0 release and I'm currently working on 4.2.0, which mainly includes group rolls #88.

My current thinking is to try and compare modifiers with compare points against the min/max values of the dice, so it throws an error if the comparison will always pass:

```
d6!>0
d6!<7

// Fudge dice have a min of -1 and max of +1
dF!>-2
dF!<2
```

And on top of that, have a maximum iteration count, that cuts it off if it loops too many times.
An asynchronous option is definitely something that could be considered.

I'm also always happy to receive PRs if you're up for taking a crack at it.

**Comment by GreenImp** [comment]
After a bit of thought, due to the amount of modifiers, and combinations of such, I think it's going to be very difficult to check the modifiers to ensure that they're sensible (e.g. will not always explode).

So I think a max cap on modifier iterations is probably the best way of solving this, perhaps if each modifier has a maximum iteration of 1000.

I'm also keen to look into the option of running it asynchronously, but will most likely look at that as a separate ticket.

**Comment by GreenImp** [comment]
@Yernemm just thought it worth saying that your ban list is probably catching a lot of false positives there:
* `'>', '<'`: These are used in quite a few modifiers, such as target success / failure, and shouldn't cause any issues in these cases. They'll only be problematic when used with exploding or re-rolling dice, as they're the only modifiers that create loops.
* `'!', "r"`: These are definitely the culprits to block if you're worried about infinite loops
* `"exp", "log", "max", "min", "pow", "sign"`: These use the standard JS math functions, so shouldn't cause any issues.

At present the only modifiers that should be causing issues are the exploding (Including compounding and penetrating), and re-rolling dice.
As mentioned above, I'm getting a fix in shortly so modifiers will only loop a maximum of 1000 times per roll, so that'll stop infinite loops from occurring entirely. So you may or may not be okay to not ban any of the modifiers.

**Comment by TapuCosmo** [comment]
Can you also add a configurable global (for all rolls for a given input) roll limit to prevent a large, but non-infinite, number of rolls from being done? For example, inputs like `(99^99)d20` would block the thread for basically forever (over 10^197 rolls).

**Comment by GreenImp** [comment]
So, there's some _kind of_ good news in disguise, although it's caused by a newly found bug.

The die quantities are all parsed to integers, using JS `parseInt()`. But `parseInt()` has [issues with very large numbers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#Description):

> Because some numbers use the e character in their string representation (e.g. 6.022e23 for 6.022 × 1023), using parseInt to truncate numbers will produce unexpected results when used on very large or very small numbers.

Running `(99^99)d20`, will calculate `(99^99)` as `3.6972963764973e+197`. When this runs through `parseInt()` it comes out as `3`.

So rolling `(99^99)d20` will actually roll `3d20`.

Obviously it needs to handle large numbers better - I've raised #159 to handle that.

But I'm thinking of limiting the die quantity to 5 digits / 99999.
After running some tests, that seems like a good max - any more than that and I start to see memory issues.

The library currently has no real sense of configurable options, so making this limit configurable would be quite a bit of work. I am looking into options for allowing finer configuration of things though, but it's not likely to be implemented soon.

**Comment by GreenImp** [comment]
Another thought - it's probably best to ensure that the numbers are also <= `Number.MAX_SAFE_INTEGER` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER)

**Comment by GreenImp** [comment]
The fix for limiting modifier iterations to 1000 is done and will be going out in v4.3.0 shortly.

I'm going to leave this ticket open though, as there's still more that can be done to reduce the chance of bringing down a server.

**Comment by GreenImp** [comment]
I've just merged in #176 which adds proper handling of very large numbers, so they don't get incorrectly parsed to integers.

You can read my full explanation there: https://github.com/GreenImp/rpg-dice-roller/issues/159#issuecomment-667235131

In short, the dice roller can now handle very large numbers, which works really well for die sides, but needs to be restricted for qty as that'll cause some memory issues. I'm going to resolve that as part of this issue.

**Comment by GreenImp** [comment]
And I've now just merged in #177, which restricts the qty of each die to `999`. After testing, I found more than that to be too heavy when dealing with lots of exploding / re-rolling dice, and also multiple dice.

So, these examples are valid:

```
999d20
999d10 + 999d6 * 999d20
```

But more than `999` per die is invalid;

```
1000d20
```

Will throw:

> RangeError: qty must be between 1 and 999

These fixes, along with the previous one to stop infinite loops should cover most bases.
I'm not certain what else there is to do on this, but I'm open to suggestions!
